File: tradebot/config/file.py
from __future__ import annotations

import json
from pathlib import Path

from pydantic import BaseModel, ValidationError

from tradebot.config.models import (
    AppConfig,
    RiskConfig,
    WatchlistConfig,
    WeightsConfig,
    WhalesConfig,
)


class DashboardConfig(BaseModel):
    enabled: bool = True
    host: str = "127.0.0.1"
    port: int = 8765


class TradeBotConfig(BaseModel):
    config_version: int = 1
    data_dir: str = "data"
    app: AppConfig
    risk: RiskConfig
    weights: WeightsConfig
    watchlist: WatchlistConfig
    dashboard: DashboardConfig
    whales: WhalesConfig = WhalesConfig()


def load_config(path: Path) -> TradeBotConfig:
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"config not found: {path}")
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        raise ValueError(f"invalid JSON in {path}: {e}") from e
    try:
        return TradeBotConfig.model_validate(data)
    except ValidationError as e:
        raise ValueError(f"config validation failed: {e}") from e


def save_config(path: Path, config: TradeBotConfig) -> None:
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(config.model_dump_json(indent=2), encoding="utf-8")

File: crates/config/src/file.rs
use crate::error::ConfigError;
use crate::models::{
    AppConfig, DashboardConfig, RiskConfig, WatchlistConfig, WeightsConfig, WhalesConfig,
};
use serde::{Deserialize, Serialize};
use std::path::Path;

fn default_config_version() -> u32 {
    1
}
fn default_data_dir() -> String {
    "data".into()
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TradeBotConfig {
    #[serde(default = "default_config_version")]
    pub config_version: u32,
    #[serde(default = "default_data_dir")]
    pub data_dir: String,
    pub app: AppConfig,
    pub risk: RiskConfig,
    pub weights: WeightsConfig,
    pub watchlist: WatchlistConfig,
    pub dashboard: DashboardConfig,
    #[serde(default)]
    pub whales: WhalesConfig,
}

impl TradeBotConfig {
    /// Run all cross-field validation. Called by `load_config` after deserialize.
    pub fn validate(&self) -> Result<(), ConfigError> {
        self.risk.validate()?;
        self.weights.validate()?;
        self.watchlist.validate()?;
        self.whales.validate()?;
        Ok(())
    }
}

/// Load and validate a config file. Mirrors `tradebot/config/file.py::load_config`.
pub fn load_config(path: &Path) -> Result<TradeBotConfig, ConfigError> {
    if !path.exists() {
        return Err(ConfigError::NotFound(path.to_path_buf()));
    }
    let text = std::fs::read_to_string(path).map_err(|source| ConfigError::Io {
        path: path.to_path_buf(),
        source,
    })?;
    let cfg: TradeBotConfig =
        serde_json::from_str(&text).map_err(|source| ConfigError::InvalidJson {
            path: path.to_path_buf(),
            source,
        })?;
    cfg.validate()?;
    Ok(cfg)
}

/// Write a config as pretty JSON. Mirrors `tradebot/config/file.py::save_config`.
pub fn save_config(path: &Path, config: &TradeBotConfig) -> Result<(), ConfigError> {
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent).map_err(|source| ConfigError::Io {
            path: parent.to_path_buf(),
            source,
        })?;
    }
    let json = serde_json::to_string_pretty(config).map_err(|source| ConfigError::InvalidJson {
        path: path.to_path_buf(),
        source,
    })?;
    std::fs::write(path, json).map_err(|source| ConfigError::Io {
        path: path.to_path_buf(),
        source,
    })?;
    Ok(())
}
